File: modules/web/js/launcher/launch-manager.js
/**
 * Copyright (c) 2017, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
 *
 * WSO2 Inc. licenses this file to you under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import _ from 'lodash';
import EventChannel from 'event_channel';
import Console from 'console';
import LaunchChannel from './launch-channel';
/**
 * Launch Manager
 * @class LaunchManager
 * @extends {EventChannel}
 */
class LaunchManager extends EventChannel {
    /**
     * Creates an instance of LaunchManager.
     *
     * @memberof LaunchManager
     */
    constructor() {
        super();
        this.enable = false;
        this.channel = undefined;
        this.active = false;
    }
    /**
     *
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    runApplication(file) {
        this.channel = new LaunchChannel({ endpoint: this.endpoint, launcher: this });
        this.openConsole();
        this.channel.on('connected', _.bindKey(this, 'sendRunApplicationMessage', file));
    }
    /**
     * Run Ballerina Service
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    runService(file) {
        this.channel = new LaunchChannel({ endpoint: this.endpoint, launcher: this });
        this.openConsole();
        this.channel.on('connected', () => { this.sendRunServiceMessage(file); });
    }
    /**
     * Run Ballerina main
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    debugApplication(file) {
        this.channel = new LaunchChannel({ endpoint: this.endpoint, launcher: this });
        this.openConsole();
        this.channel.on('connected', () => { this.sendDebugApplicationMessage(file); });
    }
    /**
     * Debug ballerina Service
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    debugService(file) {
        this.channel = new LaunchChannel({ endpoint: this.endpoint, launcher: this });
        this.openConsole();
        this.channel.on('connected', () => { this.sendDebugServiceMessage(file); });
    }
    /**
     * Send message to ballerina program
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    sendRunApplicationMessage(file) {
        const message = {
            command: 'RUN_PROGRAM',
            fileName: file.getName(),
            filePath: file.getPath(),
            commandArgs: this.getApplicationConfigs(file),
        };
        this.channel.sendMessage(message);
    }
    /**
     * Send message to ballerina service
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    sendRunServiceMessage(file) {
        const message = {
            command: 'RUN_SERVICE',
            fileName: file.getName(),
            filePath: file.getPath(),
        };
        this.channel.sendMessage(message);
    }
    /**
     * Send message to debugging ballerina program
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    sendDebugApplicationMessage(file) {
        const message = {
            command: 'DEBUG_PROGRAM',
            fileName: file.getName(),
            filePath: file.getPath(),
            commandArgs: this.getApplicationConfigs(file),
        };
        this.channel.sendMessage(message);
    }
    /**
     * Send message to debugging ballerina service
     * @param {File} file - File instance
     *
     * @memberof LaunchManager
     */
    sendDebugServiceMessage(file) {
        const message = {
            command: 'DEBUG_SERVICE',
            fileName: file.getName(),
            filePath: file.getPath(),
        };
        this.channel.sendMessage(message);
    }
    /**
     *
     * @param {Object} message - Process message from backend
     *
     * @memberof LaunchManager
     */
    processMesssage(message) {
        if (message.code === 'OUTPUT') {
            if (_.endsWith(message.message, this.debugPort)) {
                this.trigger('debug-active', this.debugEndpoint);
                return;
            }
        }
        if (message.code === 'EXECUTION_STARTED') {
            this.active = true;
            this.trigger('execution-started');
        }
        if (message.code === 'EXECUTION_STOPED' || message.code === 'EXECUTION_TERMINATED') {
            this.active = false;
            this.trigger('execution-ended');
            this.channel.close();
        }
        if (message.code === 'DEBUG_PORT') {
            this.debugPort = message.port;
            return;
        }
        if (message.code === 'EXIT') {
            this.active = false;
            this.trigger('session-ended');
            // close the current channel.
            this.channel.close();
        }
        if (message.code === 'PONG') {
            // if a pong message is received we will ignore.
            return;
        }
        if (message.code === 'INVALID_CMD') {
            // ignore and return.
            return;
        }
        Console.println(message);
    }
    /**
     * Clear and Open console view
     * @memberof LaunchManager
     */
    openConsole() {
        Console.clear();
        Console.show();
    }
    /**
     * Init launch manager
     * @param {any} options - Config options
     *
     * @memberof LaunchManager
     */
    init(options) {
        this.endpoint = _.get(options, 'application.config.services.launcher.endpoint');
        this.debugEndpoint = _.get(options, 'application.config.services.debugger.endpoint');
        this.enable = true;
        this.application = options.application;
        Console.setApplication(this.application);
    }
    /**
     * Stop ballerina program
     *
     * @memberof LaunchManager
     */
    stopProgram() {
        const message = {
            command: 'TERMINATE',
        };
        this.channel.sendMessage(message);
    }
    /**
     * Get application configs
     * @param {File} file - File instance
     * @returns {Object} application configs
     *
     * @memberof LaunchManager
     */
    getApplicationConfigs(file) {
        const args = this.application.browserStorage.get(`launcher-app-configs-${file.id}`);
        return args || '';
    }
}

export default new LaunchManager();

File: modules/services/workspace-service/src/main/java/org/ballerinalang/composer/service/workspace/launcher/LaunchManager.java
/*
 * Copyright (c) 2017, WSO2 Inc. (http://wso2.com) All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.ballerinalang.composer.service.workspace.launcher;

import com.google.gson.Gson;
import io.netty.channel.Channel;
import io.netty.handler.codec.http.websocketx.TextWebSocketFrame;
import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.StringUtils;
import org.ballerinalang.composer.service.workspace.launcher.dto.CommandDTO;
import org.ballerinalang.composer.service.workspace.launcher.dto.MessageDTO;
import org.ballerinalang.composer.service.workspace.launcher.util.LaunchUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.Charset;

/**
 * Launch Manager which manage launch requests from the clients.
 */
public class LaunchManager {

    private static final Logger logger = LoggerFactory.getLogger(LaunchManager.class);

    private static LaunchManager launchManagerInstance;

    private LaunchServer launchServer;

    private LaunchSession launchSession;

    private Command command;

    private String startedServiceURL;
    
    private String port;

    /**
     * Instantiates a new Debug manager.
     */
    protected LaunchManager() {

    }

    /**
     * Launch manager singleton
     *
     * @return LaunchManager instance
     */
    public static LaunchManager getInstance() {
        synchronized (LaunchManager.class) {
            if (launchManagerInstance == null) {
                launchManagerInstance = new LaunchManager();
            }
        }
        return launchManagerInstance;
    }


    public void init(int port, String startedServiceURL) {
        // start the debug server if it is not started yet.
        if (this.launchServer == null) {
            this.launchServer = new LaunchServer(port);
            this.launchServer.startServer();
        }

        // set URL to invoke the started service.
        this.startedServiceURL = startedServiceURL;
    }

    private void run(Command command) {
        Process program = null;
        this.command = command;
        // send a message if ballerina home is not set
        if (null == System.getProperty("ballerina.home") || System.getProperty("ballerina.home").isEmpty()) {
            pushMessageToClient(launchSession, LauncherConstants.ERROR, LauncherConstants.ERROR, LauncherConstants
                    .INVALID_BAL_PATH_MESSAGE);
            pushMessageToClient(launchSession, LauncherConstants.ERROR, LauncherConstants.ERROR, LauncherConstants
                    .SET_BAL_PATH_MESSAGE);
            return;
        }

        try {
            String cmd = command.toString();
            if (command.getPackageDir() == null) {
                program = Runtime.getRuntime().exec(cmd);
            } else {
                program = Runtime.getRuntime().exec(cmd, null, new File(command.getPackageDir()));
            }

            command.setProgram(program);


            if (command.getType() == LauncherConstants.ProgramType.RUN) {
                pushMessageToClient(launchSession, LauncherConstants.EXECUTION_STARTED, LauncherConstants.INFO,
                        String.format(LauncherConstants.RUN_MESSAGE, command.getFileName()));
            } else {
                pushMessageToClient(launchSession, LauncherConstants.EXECUTION_STARTED, LauncherConstants.INFO,
                        String.format(LauncherConstants.SERVICE_MESSAGE, command.getFileName()));
            }

            if (command.isDebug()) {
                MessageDTO debugMessage = new MessageDTO();
                debugMessage.setCode(LauncherConstants.DEBUG);
                debugMessage.setPort(command.getPort());
                pushMessageToClient(launchSession, debugMessage);
            }

            // start a new thread to stream command output.
            Runnable output = new Runnable() {
                public void run() {
                    LaunchManager.this.streamOutput();
                }
            };
            (new Thread(output)).start();
            Runnable error = new Runnable() {
                public void run() {
                    LaunchManager.this.streamError();
                }
            };
            (new Thread(error)).start();

        } catch (IOException e) {
            pushMessageToClient(launchSession, LauncherConstants.EXIT, LauncherConstants.ERROR, e.getMessage());
        }
    }

    public void streamOutput() {
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new InputStreamReader(this.command.getProgram().getInputStream(), Charset
                    .defaultCharset()));
            String line = "";
            while ((line = reader.readLine()) != null) {
                // improve "server connector started" log message to have the service URL in it.
                // This is to handle the cloud use case.
                if (line.startsWith(LauncherConstants.SERVER_CONNECTOR_STARTED_AT_HTTP_DEFAULT_PORT_LOG)
                        && startedServiceURL != null) {
                    line = LauncherConstants.SERVER_CONNECTOR_STARTED_LOG + " " + startedServiceURL;
                }

                // This is to handle local service run use case.
                if (line.startsWith(LauncherConstants.SERVER_CONNECTOR_STARTED_AT_HTTP_DEFAULT_PORT_LOG)
                        && startedServiceURL == null) {
                    line = LauncherConstants.SERVER_CONNECTOR_STARTED_LOG + " " +
                            String.format(LauncherConstants.LOCAL_TRY_IT_URL, LauncherConstants.LOCALHOST,
                                    this.getUpdatedPort(line));
                    pushMessageToClient(launchSession, LauncherConstants.OUTPUT, LauncherConstants.DATA, line);
                } else {
                    pushMessageToClient(launchSession, LauncherConstants.OUTPUT, LauncherConstants.DATA, line);
                }

            }
            pushMessageToClient(launchSession, LauncherConstants.EXECUTION_STOPED, LauncherConstants.INFO,
                    LauncherConstants.END_MESSAGE);
        } catch (IOException e) {
            logger.error("Error while sending output stream to client.", e);
        } finally {
            if (reader != null) {
                IOUtils.closeQuietly(reader);
            }
        }
    }

    public void streamError() {
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new InputStreamReader(
                    this.command.getProgram().getErrorStream(), Charset.defaultCharset()));
            String line = "";
            while ((line = reader.readLine()) != null) {
                if (this.command.isErrorOutputEnabled()) {
                    pushMessageToClient(launchSession, LauncherConstants.OUTPUT, LauncherConstants.ERROR, line);
                }
            }
        } catch (IOException e) {
            logger.error("Error while sending error stream to client.", e);
        } finally {
            if (reader != null) {
                IOUtils.closeQuietly(reader);
            }
        }
    }

    /**
     * Stop a running ballerina program
     */
    public void stopProcess() {
        int pid = -1;
        if (this.command != null && this.command.getProgram().isAlive()) {
            //shutdown error streaming to prevent kill message displaying to user.
            this.command.setErrorOutputEnabled(false);

            String os = getOperatingSystem();
            if (os == null) {
                logger.error("unsupported operating system");
                pushMessageToClient(launchSession, LauncherConstants.UNSUPPORTED_OPERATING_SYSTEM,
                        LauncherConstants.ERROR, LauncherConstants.TERMINATE_MESSAGE);
                return;
            }
            Terminator terminator = new TerminatorFactory().getTerminator(os, this.command);
            if (terminator == null) {
                logger.error("unsupported operating system");
                pushMessageToClient(launchSession, LauncherConstants.UNSUPPORTED_OPERATING_SYSTEM,
                        LauncherConstants.ERROR, LauncherConstants.TERMINATE_MESSAGE);
                return;
            }

            terminator.terminate();
            pushMessageToClient(launchSession, LauncherConstants.EXECUTION_TERMINATED, LauncherConstants.INFO,
                    LauncherConstants.TERMINATE_MESSAGE);
        }
    }

    /**
     * Returns name of the operating system running. null if not a unsupported operating system.
     * @return operating system
     */
    private String getOperatingSystem() {
        if (LaunchUtils.isWindows()) {
            return "windows";
        } else if (LaunchUtils.isMac() || LaunchUtils.isUnix() || LaunchUtils.isSolaris()) {
            return "unix";
        }
        return null;
    }

    public void addLaunchSession(Channel channel) {
        this.launchSession = new LaunchSession(channel);
    }

    public void processCommand(String json) {
        Gson gson = new Gson();
        CommandDTO command = gson.fromJson(json, CommandDTO.class);
        MessageDTO message;
        switch (command.getCommand()) {
            case LauncherConstants.RUN_PROGRAM:
                run(new Command(LauncherConstants.ProgramType.RUN, command.getFileName(), command.getFilePath(),
                        command.getCommandArgs(), false));
                break;
            case LauncherConstants.RUN_SERVICE:
                run(new Command(LauncherConstants.ProgramType.SERVICE, command.getFileName(), command.getFilePath(),
                        false));
                break;
            case LauncherConstants.DEBUG_PROGRAM:
                run(new Command(LauncherConstants.ProgramType.RUN, command.getFileName(), command.getFilePath(),
                        command.getCommandArgs(), true));
                break;
            case LauncherConstants.DEBUG_SERVICE:
                run(new Command(LauncherConstants.ProgramType.SERVICE, command.getFileName(), command.getFilePath(),
                        true));
                break;
            case LauncherConstants.TERMINATE:
                stopProcess();
                break;
            case LauncherConstants.PING:
                message = new MessageDTO();
                message.setCode(LauncherConstants.PONG);
                launchServer.pushMessageToClient(launchSession, message);
                break;
            default:
                message = new MessageDTO();
                message.setCode(LauncherConstants.INVALID_CMD);
                message.setMessage(LauncherConstants.MSG_INVALID);
                launchServer.pushMessageToClient(launchSession, message);
        }
    }

    /**
     * Push message to client.
     *
     * @param session the debug session
     * @param status  the status
     */
    public void pushMessageToClient(LaunchSession session, MessageDTO status) {
        Gson gson = new Gson();
        String json = gson.toJson(status);
        session.getChannel().write(new TextWebSocketFrame(json));
        session.getChannel().flush();
    }

    public void pushMessageToClient(LaunchSession session, String code, String type, String text) {
        MessageDTO message = new MessageDTO();
        message.setCode(code);
        message.setType(type);
        message.setMessage(text);
        pushMessageToClient(session, message);
    }
    
    /**
     * Gets the port of the from console log that starts with
     * LauncherConstants.SERVER_CONNECTOR_STARTED_AT_HTTP_DEFAULT_PORT_LOG
     * @param line The log line.
     * @return The port. 
     */
    private String getUpdatedPort(String line) {
        String hostPort = StringUtils.substringAfterLast(line,
                                            LauncherConstants.SERVER_CONNECTOR_STARTED_AT_HTTP_DEFAULT_PORT_LOG).trim();
        String port = StringUtils.substringAfterLast(hostPort, "-");
        if (StringUtils.isNotBlank(port)) {
            this.port = port;
        } else {
            this.port = "9090";
        }
        return this.port;
    }
    
    /**
     * Getter for running port.
     * @return The port.
     */
    public String getPort() {
        return this.port;
    }
}
